File: test/src/test/java/cz/vhromada/common/test/service/AbstractMovableServiceTest.java
package cz.vhromada.common.test.service;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.SoftAssertions.assertSoftly;

import cz.vhromada.common.Movable;
import cz.vhromada.common.service.AbstractMovableService;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.stub.AbstractMovableServiceStub;
import cz.vhromada.common.test.stub.MovableStub;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.cache.Cache;
import org.springframework.data.jpa.repository.JpaRepository;

/**
 * A class represents test for class {@link AbstractMovableService}.
 *
 * @author Vladimir Hromada
 */
class AbstractMovableServiceTest extends MovableServiceTest<Movable> {

    /**
     * Instance of {@link JpaRepository}
     */
    @Mock
    private JpaRepository<Movable, Integer> repository;

    /**
     * Test method for {@link AbstractMovableService#AbstractMovableService(JpaRepository, Cache, String)} with null repository.
     */
    @Test
    void constructor_NullMovableRepository() {
        assertThatThrownBy(() -> new AbstractMovableServiceStub(null, getCache(), getCacheKey(), this::getCopyItem))
            .isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link AbstractMovableService#AbstractMovableService(JpaRepository, Cache, String)} with null cache for data.
     */
    @Test
    void constructor_NullCache() {
        assertThatThrownBy(() -> new AbstractMovableServiceStub(repository, null, getCacheKey(), this::getCopyItem))
            .isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link AbstractMovableService#AbstractMovableService(JpaRepository, Cache, String)} with null cache key.
     */
    @Test
    void constructor_NullCacheKey() {
        assertThatThrownBy(() -> new AbstractMovableServiceStub(repository, getCache(), null, this::getCopyItem)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected JpaRepository<Movable, Integer> getRepository() {
        return repository;
    }

    @Override
    protected MovableService<Movable> getMovableService() {
        return new AbstractMovableServiceStub(repository, getCache(), getCacheKey(), this::getCopyItem);
    }

    @Override
    protected String getCacheKey() {
        return "data";
    }

    @Override
    protected Movable getItem1() {
        return new MovableStub(1, 0);
    }

    @Override
    protected Movable getItem2() {
        return new MovableStub(2, 1);
    }

    @Override
    protected Movable getAddItem() {
        return new MovableStub(null, 4);
    }

    @Override
    protected Movable getCopyItem() {
        return new MovableStub(10, 10);
    }

    @Override
    protected Class<Movable> getItemClass() {
        return Movable.class;
    }

    @Override
    protected void assertDataDeepEquals(final Movable expected, final Movable actual) {
        assertSoftly(softly -> {
            softly.assertThat(expected).isNotNull();
            softly.assertThat(actual).isNotNull();
        });
        assertSoftly(softly -> {
            softly.assertThat(actual.getId()).isEqualTo(expected.getId());
            softly.assertThat(actual.getPosition()).isEqualTo(expected.getPosition());
        });
    }

}

File: common-test/src/test/kotlin/cz/vhromada/common/test/service/AbstractMovableServiceTest.kt
package cz.vhromada.common.test.service

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.common.domain.AuditEntity
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.AbstractMovableService
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.test.stub.AbstractMovableServiceStub
import cz.vhromada.common.test.stub.AuditEntityStub
import cz.vhromada.common.test.utils.TestConstants
import org.assertj.core.api.SoftAssertions.assertSoftly
import org.mockito.Mock
import org.springframework.data.jpa.repository.JpaRepository

/**
 * A class represents test for class [AbstractMovableService].
 *
 * @author Vladimir Hromada
 */
class AbstractMovableServiceTest : MovableServiceTest<AuditEntity>() {

    /**
     * Instance of [JpaRepository]
     */
    @Mock
    private lateinit var repository: JpaRepository<AuditEntity, Int>

    /**
     * Instance of [AccountProvider]
     */
    @Mock
    private lateinit var accountProvider: AccountProvider

    /**
     * Instance of [TimeProvider]
     */
    @Mock
    private lateinit var timeProvider: TimeProvider

    /**
     * Cache key
     */
    private val cacheKey = "data"

    override fun getRepository(): JpaRepository<AuditEntity, Int> {
        return repository
    }

    override fun getAccountProvider(): AccountProvider {
        return accountProvider
    }

    override fun getTimeProvider(): TimeProvider {
        return timeProvider
    }

    override fun getService(): MovableService<AuditEntity> {
        return AbstractMovableServiceStub(repository, accountProvider, timeProvider, cache, cacheKey, { getCopyItem() }, { repository.findAll() })
    }

    override fun getCacheKey(): String {
        return cacheKey + TestConstants.ACCOUNT_ID
    }

    override fun getItem1(): AuditEntity {
        return AuditEntityStub(1, 0)
    }

    override fun getItem2(): AuditEntity {
        return AuditEntityStub(2, 1)
    }

    override fun getAddItem(): AuditEntity {
        return AuditEntityStub(null, 4)
    }

    override fun getCopyItem(): AuditEntity {
        return AuditEntityStub(10, 10)
    }

    override fun initAllDataMock(data: List<AuditEntity>) {
        whenever(repository.findAll()).thenReturn(data)
    }

    override fun verifyAllDataMock() {
        verify(repository).findAll()
        verifyNoMoreInteractions(repository)
    }

    override fun anyItem(): AuditEntity {
        return any()
    }

    override fun argumentCaptorItem(): KArgumentCaptor<AuditEntity> {
        return argumentCaptor()
    }

    override fun assertDataDeepEquals(expected: AuditEntity, actual: AuditEntity) {
        assertSoftly { softly ->
            softly.assertThat(expected).isNotNull
            softly.assertThat(actual).isNotNull()
        }
        assertSoftly { softly ->
            softly.assertThat(actual.id).isEqualTo(expected.id)
            softly.assertThat(actual.position).isEqualTo(expected.position)
        }
    }

}
